compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5291,18 +5291,39 @@ private void visitCheckAndCheckPanicExpr(BLangCheckedExpr checkedExpr) {
         String operatorType = checkedExpr.getKind() == NodeKind.CHECK_EXPR ? "check" : "checkpanic";
         BLangExpression exprWithCheckingKeyword = checkedExpr.expr;
         boolean firstVisit = exprWithCheckingKeyword.getBType() == null;
-        BType typeOfExprWithCheckingKeyword;
+
+        BType checkExprCandidateType;
         if (expType == symTable.noType) {
-            typeOfExprWithCheckingKeyword = symTable.noType;
+            checkExprCandidateType = symTable.noType;
         } else {
-            typeOfExprWithCheckingKeyword = BUnionType.create(null, expType, symTable.errorType);
+            boolean prevNonErrorLoggingCheck = this.nonErrorLoggingCheck;
+            this.nonErrorLoggingCheck = true;
+            int prevErrorCount = this.dlog.errorCount();
+            this.dlog.resetErrorCount();
+            this.dlog.mute();
+
+            checkedExpr.expr.cloneAttempt++;
+            BLangExpression clone = nodeCloner.cloneNode(checkedExpr.expr);
+            BType exprType = checkExpr(clone, env, expType);
+
+            this.nonErrorLoggingCheck = prevNonErrorLoggingCheck;
+            this.dlog.setErrorCount(prevErrorCount);
+            if (!prevNonErrorLoggingCheck) {
+                this.dlog.unmute();
+            }
```

Comment:
I was wondering whether we can replace this part by `BType getCandidateType(BLangCheckedExpr checkedExpr, BType checkExprCandidateType)`. Can `expType` never be `semanticError`, before we come to this place? If so, we can replace with this method